Development data script fails

The script for fetching development data appears to no longer be working. It gets so far (and does fetch some data), then hits an error:

```
$  ./util/docker-dev/dev_data.sh
...
2019/11/17 19:27:11 populate_dev_data.go:334: Added 3 TestRun entities
2019/11/17 19:27:11 fetch_runs.go:36: Fetching https://wpt.fyi/api/interop?aligned=true&label=master&product=chrome%5Bexperimental%5D&product=edge%5Bstable%5D&product=firefox%5Bexperimental%5D&product=safari%5Bexperimental%5D...
2019/11/17 19:27:12 populate_dev_data.go:251: Failed to fetch interop (?aligned=true).
2019/11/17 19:27:12 populate_dev_data.go:217: Successfully copied a total of 133 distinct TestRuns
2019/11/17 19:27:12 populate_dev_data.go:219: Adding latest production PendingTestRun...
2019/11/17 19:27:12 fetch_runs.go:36: Fetching https://wpt.fyi/api/status...
2019/11/17 19:27:41 populate_dev_data.go:332: Failed to add PendingTestRun entities: unexpected response: exception: "cgoogle.appengine.runtime.apiproxy_errors\nRequestTooLargeError\np0\n(S'The request to API call datastore_v3.Put() was too large.'\np1\ntp2\nRp3\n."
exit status 1
make: *** [dev_data] Error 1
Makefile:230: recipe for target 'dev_data' failed
```

--- a/api/pending_test_runs.go
+++ b/api/pending_test_runs.go
@@ -26,14 +26,19 @@ func apiPendingTestRunsHandler(w http.ResponseWriter, r *http.Request) {
 	case "pending":
 		q = q.Order("-Stage").Filter("Stage < ", shared.StageValid)
 	case "invalid":
-		q = q.Order("-Stage").Filter("Stage > ", shared.StageValid)
+		q = q.Filter("Stage = ", shared.StageInvalid)
+	case "empty":
+		q = q.Filter("Stage = ", shared.StageEmpty)
+	case "duplicate":
+		q = q.Filter("Stage = ", shared.StageDuplicate)
 	case "":
 		// No-op
 	default:
 		http.Error(w, "Invalid filter: "+filter, http.StatusBadRequest)
 		return
 	}
-	q = q.Order("-Updated")
+	// TODO(Hexcles): Support pagination.
+	q = q.Order("-Updated").Limit(shared.MaxCountMaxValue)
 
 	var runs []shared.PendingTestRun
 	keys, err := store.GetAll(q, &runs)

--- a/api/pending_test_runs_medium_test.go
+++ b/api/pending_test_runs_medium_test.go
@@ -3,6 +3,7 @@
 package api
 
 import (
+	"context"
 	"encoding/json"
 	"io/ioutil"
 	"net/http"
@@ -18,6 +19,13 @@ import (
 	"google.golang.org/appengine/datastore"
 )
 
+func createPendingRun(ctx context.Context, run *shared.PendingTestRun) error {
+	key := datastore.NewIncompleteKey(ctx, "PendingTestRun", nil)
+	key, err := datastore.Put(ctx, key, run)
+	run.ID = key.IntID()
+	return err
+}
+
 func TestAPIPendingTestHandler(t *testing.T) {
 	i, err := sharedtest.NewAEInstance(true)
 	assert.Nil(t, err)
@@ -30,23 +38,40 @@ func TestAPIPendingTestHandler(t *testing.T) {
 	now := time.Now().Truncate(time.Minute).In(time.UTC)
 	yesterday := now.Add(time.Hour * -24)
 
-	invalid := shared.PendingTestRun{}
-	invalid.Created = yesterday
-	invalid.Updated = now
-	invalid.Stage = shared.StageInvalid
-	key := datastore.NewIncompleteKey(ctx, "PendingTestRun", nil)
-	key, err = datastore.Put(ctx, key, &invalid)
-	assert.Nil(t, err)
-	invalid.ID = key.IntID()
+	invalid := shared.PendingTestRun{
+		Created: yesterday,
+		Updated: now,
+		Stage:   shared.StageInvalid,
+	}
+	assert.Nil(t, createPendingRun(ctx, &invalid))
 
-	running := shared.PendingTestRun{}
-	running.Created = yesterday.Add(time.Hour)
-	running.Updated = now.Add(time.Minute * -5)
-	running.Stage = shared.StageCIRunning
-	key = datastore.NewIncompleteKey(ctx, "PendingTestRun", nil)
-	key, err = datastore.Put(ctx, key, &running)
-	assert.Nil(t, err)
-	running.ID = key.IntID()
+	empty := shared.PendingTestRun{
+		Created: yesterday,
+		Updated: now.Add(time.Minute * -1),
+		Stage:   shared.StageEmpty,
+	}
+	assert.Nil(t, createPendingRun(ctx, &empty))
+
+	duplicate := shared.PendingTestRun{
+		Created: yesterday,
+		Updated: now.Add(time.Minute * -2),
+		Stage:   shared.StageDuplicate,
+	}
+	assert.Nil(t, createPendingRun(ctx, &duplicate))
+
+	received := shared.PendingTestRun{
+		Created: yesterday.Add(time.Hour),
+		Updated: now.Add(time.Minute * -10),
+		Stage:   shared.StageWptFyiReceived,
+	}
+	assert.Nil(t, createPendingRun(ctx, &received))
+
+	running := shared.PendingTestRun{
+		Created: yesterday.Add(time.Hour),
+		Updated: now.Add(time.Minute * -5),
+		Stage:   shared.StageCIRunning,
+	}
+	assert.Nil(t, createPendingRun(ctx, &running))
 
 	t.Run("/api/status", func(t *testing.T) {
 		r, _ = i.NewRequest("GET", "/api/status", nil)
@@ -56,9 +81,13 @@ func TestAPIPendingTestHandler(t *testing.T) {
 		assert.Equal(t, http.StatusOK, resp.Code, string(body))
 		var results []shared.PendingTestRun
 		json.Unmarshal(body, &results)
-		assert.Len(t, results, 2)
+		assert.Len(t, results, 5)
+		// Sorted by Update.
 		assert.Equal(t, results[0].ID, invalid.ID)
-		assert.Equal(t, results[1].ID, running.ID)
+		assert.Equal(t, results[1].ID, empty.ID)
+		assert.Equal(t, results[2].ID, duplicate.ID)
+		assert.Equal(t, results[3].ID, running.ID)
+		assert.Equal(t, results[4].ID, received.ID)
 	})
 
 	t.Run("/api/status/pending", func(t *testing.T) {
@@ -70,20 +99,39 @@ func TestAPIPendingTestHandler(t *testing.T) {
 		assert.Equal(t, http.StatusOK, resp.Code, string(body))
 		var results []shared.PendingTestRun
 		json.Unmarshal(body, &results)
-		assert.Len(t, results, 1)
+		assert.Len(t, results, 2)
 		assert.Equal(t, results[0].ID, running.ID)
+		assert.Equal(t, results[1].ID, received.ID)
 	})
 
-	t.Run("/api/status/invalid", func(t *testing.T) {
-		r, _ = i.NewRequest("GET", "/api/status/invalid", nil)
-		r = mux.SetURLVars(r, map[string]string{"filter": "invalid"})
-		resp := httptest.NewRecorder()
-		apiPendingTestRunsHandler(resp, r)
-		body, _ := ioutil.ReadAll(resp.Result().Body)
-		assert.Equal(t, http.StatusOK, resp.Code, string(body))
-		var results []shared.PendingTestRun
-		json.Unmarshal(body, &results)
-		assert.Len(t, results, 1)
-		assert.Equal(t, results[0].ID, invalid.ID)
-	})
+	filters := []string{"invalid", "empty", "duplicate"}
+	runs := []*shared.PendingTestRun{&invalid, &empty, &duplicate}
+
+	for index, filter := range filters {
+		url := "/api/status/" + filter
+		t.Run(url, func(t *testing.T) {
+			r, _ = i.NewRequest("GET", url, nil)
+			r = mux.SetURLVars(r, map[string]string{"filter": filter})
+			resp := httptest.NewRecorder()
+			apiPendingTestRunsHandler(resp, r)
+			body, _ := ioutil.ReadAll(resp.Result().Body)
+			assert.Equal(t, http.StatusOK, resp.Code, string(body))
+			var results []shared.PendingTestRun
+			json.Unmarshal(body, &results)
+			assert.Len(t, results, 1)
+			assert.Equal(t, results[0].ID, runs[index].ID)
+		})
+	}
+}
+
+func TestAPIPendingTestHandler_invalidFilter(t *testing.T) {
+	i, err := sharedtest.NewAEInstance(true)
+	assert.Nil(t, err)
+	defer i.Close()
+
+	r, _ := i.NewRequest("GET", "/api/status/foobar", nil)
+	r = mux.SetURLVars(r, map[string]string{"filter": "foobar"})
+	resp := httptest.NewRecorder()
+	apiPendingTestRunsHandler(resp, r)
+	assert.Equal(t, http.StatusBadRequest, resp.Code)
 }

--- a/api/routes.go
+++ b/api/routes.go
@@ -62,7 +62,7 @@ func RegisterRoutes() {
 	pendingTestRuns := shared.WrapApplicationJSON(
 		shared.WrapPermissiveCORS(apiPendingTestRunsHandler))
 	shared.AddRoute("/api/status", "api-pending-test-runs", pendingTestRuns)
-	shared.AddRoute("/api/status/{filter:pending|invalid}", "api-pending-test-runs", pendingTestRuns)
+	shared.AddRoute("/api/status/{filter:pending|invalid|empty|duplicate}", "api-pending-test-runs", pendingTestRuns)
 
 	// API endpoint for redirecting to a run's summary JSON blob.
 	shared.AddRoute("/api/results", "api-results", shared.WrapPermissiveCORS(apiResultsRedirectHandler))

